src/wp-includes/html-api/class-wp-html-processor.php

```diff
@@ -19,22 +23,158 @@ class WP_HTML_Processor extends WP_HTML_Tag_Processor {
 	 *
 	 * @return boolean Whether an element was found.
 	 */
-	public function step( $insertion_mode = 'in-body' ) {
-		switch ( $insertion_mode ) {
+	public function step( $insertion_mode = null ) {
+		switch ( $insertion_mode ?: $this->insertion_mode ) {
 			case 'in-body':
 				return $this->step_in_body();
 
 			default:
-				return false;
+				return self::NOT_IMPLEMENTED_YET;
 		}
 	}
 
+	/**
+	 * Parses next element in the 'in head' insertion mode.
+	 *
+	 * Not yet implemented.
+	 *
+	 * @see https://html.spec.whatwg.org/#parsing-main-inhead
+	 *
+	 * @return false
+	 */
+	private function step_in_head() {
+		return self::NOT_IMPLEMENTED_YET;
+	}
+
 	/**
 	 * Parses next element in the 'in body' insertion mode.
 	 *
+	 * @see https://html.spec.whatwg.org/#parsing-main-inbody
+	 *
 	 * @return boolean Whether an element was found.
 	 */
 	private function step_in_body() {
+		ignored:
+		parent::set_bookmark( 'current' );
+		if ( ! $this->next_tag( self::$query ) ) {
```

Comment:
What use case did you have in mind for supporting `$query`? It makes me anxious about accidentally skipping over tags even though parsing HTML requires consuming all of them.

Reply:
Ah, 'visit_closers' 👍

Reply:
also I think this should be `parent::next_tag( self::$query )` 🤷‍♂️

Reply:
I wish we called that method just `next()`, subclassing would be easier. What do you think about deprecating `next_tag` and refactoring to `next`?

Reply:
don't have much thought yet. people should never call `WP_HTML_Processor::next_tag()` though